[BOLT] Optimize basic block loops to avoid n^2 loop

Mark-Simulacrum opened the pull request:
This improves BOLT runtime when optimizing rustc_driver.so from 15 minutes to 7 minutes (or 49 minutes to 37 minutes of userspace time).

Mark-Simulacrum committed:
[BOLT] Optimize basic block loops to avoid n^2 loop

This improves BOLT runtime when optimizing rustc_driver.so from 15
minutes to 7 minutes (49 minutes to 37 minutes of userspace time).

paschalis-mpeis reviewed (commented):
Hey @Mark-Simulacrum,

Thanks for your patch. Tests appear to hang, however, your current version includes 69ccc394713aca2c031b748c5020c9da2c31de53 (now reverted) which caused some issues.

Could you rebase to latest main so we can verify that?

Mark-Simulacrum commented:
Rebased. Can you clarify how you're running tests, so I can iterate locally as well?

paschalis-mpeis reviewed (commented):
Hey Mark,

Thanks for rebasing. I can now verify that tests complete.

There are a few more users of `getBasicBlockAfter` in loops (the call you cached) that might also benefit from this.
I am not sure if it's worth wrapping this in a helper that returns the cache, possibly parameterized by `IgnoreSplits` as well. Then others can re-use.

You can use the `ninja check-bolt` target to run the unit / lit tests.

Mark-Simulacrum committed:
[BOLT] Split getBasicBlocksAfter cache into a distinct function

This enables future re-use in other code that calls getBasicBlockAfter
in loops, though for now those uses aren't introduced.

maksfb commented:
Thanks for the patch. The time it takes to run BOLT sounds excessive, even 7 minutes is a lot. Could you share BOLT log with `--time-rewrite --time-opts`?

Kobzol commented:
The size of `.text.warm` back then was `841 250` B (~840 KiB). The absolute max RSS wins are shown [here](https://perf.rust-lang.org/compare.html?start=774ae599aba35f7b323e77be88923b8f476ee783&end=9023f908cfbe7a475f369717a61cb8eb865cfd25&stat=max-rss&showRawData=true). They are in the range of ~2 MiB to 30 MiB.

maksfb commented:
Thanks for checking.

30 MiB is hard to explain, unless it's a number gathered across multiple processes, e.g. in a case of a parallel build. If it's the number for a single process, it must be some side effect. I find it hard to believe it's a direct impact from a code layout optimization.

Kobzol commented:
We haven't tried this exact combination, I'll try it, thanks. The issue we had with `--use-old-text` was that it was "unstable", i.e. in one commit the old text segment was reused, and in the next it wasn't, so the binary size was jumping all over the place, so we just disabled `--use-old-text`.

aaupov commented:
> I suppose that this PR can be closed, now that the change has landed in #160407?

Yes, the quadratic behavior in cdsplit is addressed by #160407
